spring-kafka/src/main/java/org/springframework/kafka/core/KafkaTemplate.java

```diff
@@ -283,8 +329,23 @@ public void onCompletion(RecordMetadata metadata, Exception exception) {
 		return future;
 	}
 
+	protected boolean inTransaction() {
+		return this.producers.get() != null || TransactionSynchronizationManager.isActualTransactionActive();
+	}
+
 	private Producer<K, V> getTheProducer() {
-		return this.producerFactory.createProducer();
+		if (this.transactional) {
+			Producer<K, V> producer = this.producers.get();
+			if (producer != null) {
+				return producer;
+			}
+			KafkaResourceHolder<K, V> holder = ProducerFactoryUtils
+					.getTransactionalResourceHolder(this.producerFactory);
```

Comment:
I don't think we need to add that complication - the user can provide a separate producer factory and template for non-transactional work.